File: server-pedidos-java/src/main/java/br/edu/utfpr/serverpedidosjava/controller/dto/ProdutoDTO.java
package br.edu.utfpr.serverpedidosjava.controller.dto;

import br.edu.utfpr.serverpedidosjava.entity.Produto;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;
import jakarta.validation.constraints.Size;

import java.math.BigDecimal;

public record ProdutoDTO(
        Integer id,
        @NotBlank(message = "{nome.notblank}")
        @Size(max = 100, message = "{nome.size}")
        String nome,
        @Positive(message = "{preco.positive}")
        BigDecimal preco,
        @NotNull(message = "{categoria.notnull}")
        CategoriaDTO categoria
) {
    public Produto toEntity() {
        Produto produto = new Produto();
        produto.setId(this.id);
        produto.setNome(this.nome);
        produto.setPreco(this.preco);
        produto.setCategoria(this.categoria.toEntity());
        return produto;
    }

    public static ProdutoDTO fromEntity(Produto produto) {
        return new ProdutoDTO(
                produto.getId(),
                produto.getNome(),
                produto.getPreco(),
                CategoriaDTO.fromEntity(produto.getCategoria())
        );
    }
}

File: server-pedidos/src/main/kotlin/br/edu/utfpr/serverpedidos/controller/dto/ProdutoDTO.kt
package br.edu.utfpr.serverpedidos.controller.dto

import br.edu.utfpr.serverpedidos.entity.Produto
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.NotNull
import jakarta.validation.constraints.Positive
import jakarta.validation.constraints.Size
import java.math.BigDecimal

data class ProdutoDTO(
    val id: Int,
    @field:NotBlank(message = "{nome.notblank}")
    @field:Size(max = 100, message = "{nome.size}")
    val nome: String = "",
    @field:Positive(message = "{preco.positive}")
    val preco: BigDecimal = BigDecimal.ZERO,
    @field:NotNull(message = "{categoria.notnull}")
    val categoria: CategoriaDTO? = null
) {
    fun toEntity(): Produto = Produto(
        id = this.id,
        nome = this.nome,
        preco = this.preco,
        categoria = this.categoria?.toEntity()
    )

    companion object {
        fun fromEntity(produto: Produto): ProdutoDTO = ProdutoDTO(
            id = produto.id,
            nome = produto.nome,
            preco = produto.preco,
            categoria = produto.categoria?.let { CategoriaDTO.fromEntity(it) }
        )
    }
}
